File: frontend/src/types/governance.ts
export interface GovernancePolicyItem {
  id: string;
  name: string;
  description?: string | null;
  category: "Security" | "FinOps" | "SRE" | "Kubernetes" | "Tagging" | "Operations" | string;
  severity: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" | string;
  provider: "AWS" | "Azure" | "GCP" | "Kubernetes" | "Multi-Cloud" | string;
  resource_type: string;
  rule_identifier: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface GovernancePolicyCreatePayload {
  name: string;
  description?: string;
  category?: string;
  severity?: string;
  provider?: string;
  resource_type?: string;
  rule_identifier?: string;
  enabled?: boolean;
}

export interface GovernancePolicyListResponse {
  policies: GovernancePolicyItem[];
  total: number;
}

export interface PolicyEvaluationItem {
  policy_name: string;
  rule_identifier: string;
  category: string;
  severity: string;
  provider: string;
  resource_id: string;
  resource_name: string;
  resource_type: string;
  region: string;
  status: "PASS" | "FAIL" | "WARNING" | "NOT_APPLICABLE" | string;
  evidence: string;
  recommended_action: string;
  evaluated_at: string;
}

export interface PolicyEvaluationListResponse {
  evaluations: PolicyEvaluationItem[];
  total: number;
}

export interface ComplianceFrameworkItem {
  framework: string;
  version: string;
  disclaimer: string;
  total_controls: number;
  passing_controls: number;
  failing_controls: number;
  coverage_percentage: number;
  compliance_score: number;
  status: "PASS" | "WARNING" | "FAIL" | string;
}

export interface ComplianceFrameworkListResponse {
  frameworks: ComplianceFrameworkItem[];
  total: number;
}

export interface GovernanceViolationItem {
  id: string;
  policy_id: string;
  policy_name: string;
  category: string;
  severity: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" | string;
  provider: string;
  resource_id: string;
  resource_name: string;
  resource_type: string;
  region: string;
  status: "OPEN" | "ACKNOWLEDGED" | "IN_REMEDIATION" | "RESOLVED" | "WAIVED" | string;
  evidence: string;
  recommended_action: string;
  waived_by?: string | null;
  waiver_reason?: string | null;
  detected_at: string;
  updated_at: string;
}

export interface GovernanceViolationListResponse {
  violations: GovernanceViolationItem[];
  total_violations: number;
  critical_violations: number;
}

export interface GovernanceRemediationItem {
  id: string;
  violation_id: string;
  resource: string;
  category: string;
  severity: string;
  reason: string;
  evidence: string;
  recommended_action: string;
  estimated_effort: string;
  risk_reduction: string;
  confidence: number;
  workflow_automation_supported: boolean;
}

export interface GovernanceRemediationListResponse {
  remediations: GovernanceRemediationItem[];
  total: number;
}

export interface AuditEventItem {
  id: string;
  action: string;
  actor_user_id?: string | null;
  details: Record<string, any>;
  timestamp: string;
}

export interface AuditTrailListResponse {
  audit_events: AuditEventItem[];
  total: number;
}

export interface GovernanceTrendPoint {
  day: string;
  score: number;
  violations: number;
}

export interface GovernanceTrendResponse {
  horizon_days: number;
  compliance_trend: GovernanceTrendPoint[];
  resolved_violations_period: number;
  new_violations_period: number;
  policy_coverage_percentage: number;
}

export interface GovernanceOverviewResponse {
  governance_score: number;
  governance_rating: "EXCELLENT" | "GOOD" | "AT_RISK" | "CRITICAL" | string;
  compliance_score: number;
  policies_evaluated_count: number;
  passing_controls_count: number;
  failing_controls_count: number;
  open_violations: number;
  critical_violations: number;
  high_violations: number;
  medium_violations: number;
  low_violations: number;
  data_source?: string;
  scoring_methodology?: string;
}

export interface GovernanceAnalyzeResponse {
  executive_summary: string;
  critical_violations: string[];
  framework_insights: string[];
  remediation_recommendations: GovernanceRemediationItem[];
  analyzed_at: string;
  analysis_engine?: string;
}

File: backend/app/schemas/governance.py
"""
Pydantic schemas for Enterprise Cloud Governance & Compliance Center.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

# ── Policy & Evaluation Schemas ───────────────────────────────────────────────


class GovernancePolicyItem(BaseModel):
  id: uuid.UUID
  name: str
  description: str | None = None
  category: str  # Security | FinOps | SRE | Kubernetes | Tagging | Operations
  severity: str  # LOW | MEDIUM | HIGH | CRITICAL
  provider: str  # AWS | Azure | GCP | Kubernetes | Multi-Cloud
  resource_type: str
  rule_identifier: str
  enabled: bool
  created_at: datetime
  updated_at: datetime

  model_config = {"from_attributes": True}


class GovernancePolicyCreatePayload(BaseModel):
  name: str = Field(..., min_length=3)
  description: str | None = None
  category: str = Field(default="Security")
  severity: str = Field(default="MEDIUM")
  provider: str = Field(default="Multi-Cloud")
  resource_type: str = Field(default="cloud_resource")
  rule_identifier: str | None = None
  enabled: bool = Field(default=True)


class GovernancePolicyUpdatePayload(BaseModel):
  name: str | None = None
  description: str | None = None
  category: str | None = None
  severity: str | None = None
  provider: str | None = None
  enabled: bool | None = None


class GovernancePolicyListResponse(BaseModel):
  policies: list[GovernancePolicyItem]
  total: int


class PolicyEvaluationItem(BaseModel):
  policy_name: str
  rule_identifier: str
  category: str
  severity: str
  provider: str
  resource_id: str
  resource_name: str
  resource_type: str
  region: str
  status: str  # PASS | FAIL | WARNING | NOT_APPLICABLE
  evidence: str
  recommended_action: str
  evaluated_at: str


class PolicyEvaluationListResponse(BaseModel):
  evaluations: list[PolicyEvaluationItem]
  total: int


# ── Framework & Violation Schemas ─────────────────────────────────────────────


class ComplianceFrameworkItem(BaseModel):
  framework: str  # CIS Controls | SOC 2 Type II | ISO/IEC 27001 | NIST SP 800-53 | PCI DSS
  version: str
  disclaimer: str = Field(
      default="Internal Control Mapping — Not a Certification"
  )
  total_controls: int
  passing_controls: int
  failing_controls: int
  coverage_percentage: float
  compliance_score: float
  status: str  # PASS | WARNING | FAIL


class ComplianceFrameworkListResponse(BaseModel):
  frameworks: list[ComplianceFrameworkItem]
  total: int


class GovernanceViolationItem(BaseModel):
  id: uuid.UUID
  policy_id: uuid.UUID
  policy_name: str
  category: str
  severity: str  # LOW | MEDIUM | HIGH | CRITICAL
  provider: str
  resource_id: str
  resource_name: str
  resource_type: str
  region: str
  status: str  # OPEN | ACKNOWLEDGED | IN_REMEDIATION | RESOLVED | WAIVED
  evidence: str
  recommended_action: str
  waived_by: uuid.UUID | None = None
  waiver_reason: str | None = None
  detected_at: datetime
  updated_at: datetime

  model_config = {"from_attributes": True}


class GovernanceViolationStatusPayload(BaseModel):
  status: str = Field(
      ..., description="OPEN | ACKNOWLEDGED | IN_REMEDIATION | RESOLVED | WAIVED"
  )
  reason: str | None = None


class GovernanceViolationListResponse(BaseModel):
  violations: list[GovernanceViolationItem]
  total_violations: int
  critical_violations: int


# ── Domain Governance & Remediation Schemas ───────────────────────────────────


class DomainGovernanceBreakdown(BaseModel):
  cost_governance: dict[str, Any]
  security_governance: dict[str, Any]
  sre_governance: dict[str, Any]
  kubernetes_governance: dict[str, Any]


class GovernanceRemediationItem(BaseModel):
  id: str
  violation_id: str
  resource: str
  category: str
  severity: str
  reason: str
  evidence: str
  recommended_action: str
  estimated_effort: str
  risk_reduction: str
  confidence: float
  workflow_automation_supported: bool = True


class GovernanceRemediationListResponse(BaseModel):
  remediations: list[GovernanceRemediationItem]
  total: int


# ── Audit & Trend Schemas ─────────────────────────────────────────────────────


class AuditEventItem(BaseModel):
  id: uuid.UUID
  action: str
  actor_user_id: uuid.UUID | None = None
  details: dict[str, Any]
  timestamp: datetime


class AuditTrailListResponse(BaseModel):
  audit_events: list[AuditEventItem]
  total: int


class GovernanceTrendPoint(BaseModel):
  day: str
  score: float
  violations: int


class GovernanceTrendResponse(BaseModel):
  horizon_days: int
  compliance_trend: list[GovernanceTrendPoint]
  resolved_violations_period: int
  new_violations_period: int
  policy_coverage_percentage: float


# ── Overview & AI Analysis Schemas ────────────────────────────────────────────


class GovernanceOverviewResponse(BaseModel):
  governance_score: float
  governance_rating: str  # EXCELLENT | GOOD | AT_RISK | CRITICAL
  compliance_score: float
  policies_evaluated_count: int
  passing_controls_count: int
  failing_controls_count: int
  open_violations: int
  critical_violations: int
  high_violations: int
  medium_violations: int
  low_violations: int
  data_source: str = Field(
      default="Local Governance Data — AWS/Azure/GCP/Kubernetes Fixtures"
  )
  scoring_methodology: str = Field(
      default=(
          "Weighted formula combining framework compliance score (70%),"
          " security policy coverage, cost-center tag alignment, SRE SLO"
          " governance, and Kubernetes pod security limits."
      )
  )


class GovernanceAnalyzeResponse(BaseModel):
  executive_summary: str
  critical_violations: list[str]
  framework_insights: list[str]
  remediation_recommendations: list[GovernanceRemediationItem]
  analyzed_at: str
  analysis_engine: str = Field(
      default="Local Governance Intelligence"
  )
